### ISSUE TITLE
P1.2: Add PyTorch Profiler Integration for AI Services (Selective GPU Profiling)

### ISSUE THREAD

**Issue by mikesvoboda** [mikesvoboda/nemotron-v3-home-security-intelligence#4756]
## Overview

Integrate PyTorch Profiler into AI services (yolo26, florence, clip, enrichment) to capture CUDA kernel timing, tensor operations, and GPU memory allocation. This provides deep visibility into GPU-bound inference that eBPF/CPU profiling cannot capture.

## Parent Epic

NEM-3916: Production-Grade Pyroscope Continuous Profiling Platform

## Why This Matters

Current profiling captures CPU time but AI inference is GPU-bound:

* YOLO26: TensorRT kernels, NMS post-processing
* Florence: Vision encoder, cross-attention, beam search
* CLIP: Vision transformer, embedding normalization
* Enrichment: Multi-model orchestration

Without GPU profiling, we cannot identify:

* Slow CUDA kernels
* Memory transfer bottlenecks (CPU↔GPU)
* Tensor allocation patterns
* Batch size optimization opportunities

### PyTorch Profiler Capabilities

```python
with torch.profiler.profile(
    activities=[
        torch.profiler.ProfilerActivity.CPU,
        torch.profiler.ProfilerActivity.CUDA,
    ],
    record_shapes=True,
    profile_memory=True,
    with_stack=True,
) as prof:
    model(input)

# Export to Chrome trace format (viewable in Perfetto)
prof.export_chrome_trace("trace.json")
```

### Selective Profiling Strategy

Profile only a percentage of requests to minimize overhead:

```python
import random

PROFILE_SAMPLE_RATE = float(os.getenv("PYTORCH_PROFILE_RATE", "0.05"))  # 5%

def should_profile() -> bool:
    return random.random() < PROFILE_SAMPLE_RATE
```

### Step 1: Create Shared Profiling Utility

File: `ai/shared/gpu_profiler.py` (new file)

```python
"""GPU Profiling utility for AI services.

Provides selective PyTorch profiling with export to multiple formats.
"""

import os
import random
import time
from contextlib import contextmanager
from pathlib import Path
from typing import Optional

import torch

# Configuration
PROFILE_SAMPLE_RATE = float(os.getenv("PYTORCH_PROFILE_RATE", "0.05"))
PROFILE_OUTPUT_DIR = Path(os.getenv("PYTORCH_PROFILE_DIR", "/tmp/profiles"))
PROFILE_ENABLED = os.getenv("PYTORCH_PROFILE_ENABLED", "true").lower() == "true"


def should_profile() -> bool:
    """Determine if this request should be profiled based on sample rate."""
    if not PROFILE_ENABLED:
        return False
    if not torch.cuda.is_available():
        return False
    return random.random() < PROFILE_SAMPLE_RATE


@contextmanager
def gpu_profile(
    name: str,
    trace_id: Optional[str] = None,
    record_shapes: bool = True,
    profile_memory: bool = True,
    with_stack: bool = False,  # Expensive, disable by default
):
    """Context manager for GPU profiling.
    
    Args:
        name: Profile name (e.g., "yolo26_inference")
        trace_id: Optional trace ID for correlation
        record_shapes: Record tensor shapes
        profile_memory: Track memory allocation
        with_stack: Include Python stack traces (high overhead)
    
    Usage:
        if should_profile():
            with gpu_profile("inference", trace_id=req.trace_id):
                result = model(input)
        else:
            result = model(input)
    """
    if not should_profile():
        yield None
        return
    
    PROFILE_OUTPUT_DIR.mkdir(parents=True, exist_ok=True)
    
    activities = [torch.profiler.ProfilerActivity.CPU]
    if torch.cuda.is_available():
        activities.append(torch.profiler.ProfilerActivity.CUDA)
    
    with torch.profiler.profile(
        activities=activities,
        record_shapes=record_shapes,
        profile_memory=profile_memory,
        with_stack=with_stack,
    ) as prof:
        yield prof
    
    # Generate unique filename
    timestamp = int(time.time() * 1000)
    trace_suffix = f"_{trace_id}" if trace_id else ""
    filename = f"{name}{trace_suffix}_{timestamp}.json"
    output_path = PROFILE_OUTPUT_DIR / filename
    
    # Export Chrome trace (viewable in chrome://tracing or Perfetto)
    prof.export_chrome_trace(str(output_path))
    
    # Log top operations
    print(prof.key_averages().table(sort_by="cuda_time_total", row_limit=10))


def get_gpu_memory_stats() -> dict:
    """Get current GPU memory statistics."""
    if not torch.cuda.is_available():
        return {}
    
    return {
        "allocated_mb": torch.cuda.memory_allocated() / 1024 / 1024,
        "reserved_mb": torch.cuda.memory_reserved() / 1024 / 1024,
        "max_allocated_mb": torch.cuda.max_memory_allocated() / 1024 / 1024,
    }
```

### Step 2: Integrate into YOLO26 Service

File: `ai/yolo26/model.py`

Add profiling around inference:

```python
import sys
sys.path.insert(0, str(Path(__file__).parent.parent / "shared"))
from gpu_profiler import gpu_profile, should_profile, get_gpu_memory_stats

class YOLO26Model:
    async def detect(self, image: Image.Image, trace_id: Optional[str] = None):
        # Preprocessing (CPU)
        input_tensor = self.preprocess(image)
        
        # Inference (GPU) - selectively profile
        if should_profile():
            with gpu_profile("yolo26_inference", trace_id=trace_id):
                with torch.no_grad():
                    outputs = self.model(input_tensor)
        else:
            with torch.no_grad():
                outputs = self.model(input_tensor)
        
        # Postprocessing (CPU)
        detections = self.postprocess(outputs)
        
        return detections
```

### Step 3: Integrate into Florence Service

File: `ai/florence/model.py`

```python
from gpu_profiler import gpu_profile, should_profile

class Florence2Model:
    async def generate(self, image: Image.Image, prompt: str, trace_id: Optional[str] = None):
        if should_profile():
            with gpu_profile("florence_inference", trace_id=trace_id, profile_memory=True):
                outputs = self._run_inference(image, prompt)
        else:
            outputs = self._run_inference(image, prompt)
        return outputs
```

### Step 4: Integrate into CLIP Service

File: `ai/clip/model.py`

```python
from gpu_profiler import gpu_profile, should_profile

class CLIPEmbeddingModel:
    def extract_embedding(self, image: Image.Image, trace_id: Optional[str] = None):
        if should_profile():
            with gpu_profile("clip_embedding", trace_id=trace_id):
                embedding = self._compute_embedding(image)
        else:
            embedding = self._compute_embedding(image)
        return embedding
```

### Step 5: Add Environment Variables to docker-compose

File: `docker-compose.prod.yml`

```yaml
x-ai-profiling: &ai-profiling
  PYTORCH_PROFILE_ENABLED: "true"
  PYTORCH_PROFILE_RATE: "0.05"  # 5% of requests
  PYTORCH_PROFILE_DIR: "/tmp/profiles"

services:
  ai-yolo26:
    environment:
      <<: *ai-profiling
    volumes:
      - pytorch-profiles:/tmp/profiles

  ai-florence:
    environment:
      <<: *ai-profiling
    volumes:
      - pytorch-profiles:/tmp/profiles

  ai-clip:
    environment:
      <<: *ai-profiling
    volumes:
      - pytorch-profiles:/tmp/profiles

volumes:
  pytorch-profiles:
```

### Step 6: Create Profile Viewer Endpoint (Optional)

File: `backend/api/routes/debug.py`

```python
from fastapi import APIRouter
from fastapi.responses import FileResponse
import os

router = APIRouter(prefix="/debug", tags=["debug"])

@router.get("/profiles")
async def list_profiles():
    """List available GPU profiles."""
    profile_dir = "/tmp/profiles"
    if not os.path.exists(profile_dir):
        return {"profiles": []}
    
    profiles = sorted(os.listdir(profile_dir), reverse=True)[:50]
    return {"profiles": profiles}

@router.get("/profiles/{filename}")
async def get_profile(filename: str):
    """Download a GPU profile trace file."""
    filepath = f"/tmp/profiles/{filename}"
    if not os.path.exists(filepath):
        raise HTTPException(404, "Profile not found")
    return FileResponse(filepath, media_type="application/json")
```

### Step 7: Document Usage

File: `docs/guides/profiling.md`

```markdown
## GPU Profiling with PyTorch Profiler

### Viewing Traces

1. Download a profile:
   ```bash
   curl http://localhost:8000/debug/profiles/yolo26_inference_1706400000.json -o trace.json
```

2. Open in Perfetto:
   * Navigate to [https://ui.perfetto.dev/](<https://ui.perfetto.dev/>)
   * Drag and drop the trace.json file
   * Analyze CUDA kernel timing and memory operations

### Interpreting Results

Key metrics to look for:

* **cuda_time_total**: Total GPU time per operation
* **self_cuda_time**: GPU time excluding child ops
* **cuda_memory_usage**: Memory allocated per operation

### Adjusting Sample Rate

```bash
# Profile 10% of requests
export PYTORCH_PROFILE_RATE=0.10

# Disable profiling entirely
export PYTORCH_PROFILE_ENABLED=false
```

```

## Acceptance Criteria

- [ ] `ai/shared/gpu_profiler.py` created with `gpu_profile()` context manager
- [ ] YOLO26 service integrated with selective profiling
- [ ] Florence service integrated with selective profiling
- [ ] CLIP service integrated with selective profiling
- [ ] Environment variables added to docker-compose
- [ ] Profile files written to shared volume
- [ ] Debug endpoint to list/download profiles
- [ ] Documentation with Perfetto viewing instructions
- [ ] Overhead < 5% at default 5% sample rate

## Files to Create/Modify

| File | Changes |
|------|---------|
| `ai/shared/gpu_profiler.py` | NEW: GPU profiling utility |
| `ai/yolo26/model.py` | Add profiling integration |
| `ai/florence/model.py` | Add profiling integration |
| `ai/clip/model.py` | Add profiling integration |
| `ai/enrichment/model.py` | Add profiling integration |
| `docker-compose.prod.yml` | Add env vars and volume |
| `backend/api/routes/debug.py` | Add profile viewer endpoint |
| `docs/guides/profiling.md` | Document GPU profiling |

## Performance Considerations

- 5% sample rate: ~0.25% average overhead
- 10% sample rate: ~0.5% average overhead
- `with_stack=True` adds significant overhead (10-20%), disabled by default
- Profile files are ~1-5MB each, implement rotation

## Dependencies

- PyTorch >= 2.0 (already installed in AI services)
- Shared volume for profile storage

## Blocked By

- None (can run in parallel with P0 tasks)

## Blocks

- P2.2 (GPU Observability Dashboard)

## References

- [PyTorch Profiler Documentation](https://pytorch.org/docs/stable/profiler.html)
- [Perfetto Trace Viewer](https://ui.perfetto.dev/)
- [PyTorch Profiler Tutorial](https://pytorch.org/tutorials/recipes/recipes/profiler_recipe.html)
```
